firebase_admin/app_check.py

```diff
@@ -0,0 +1,116 @@
+from typing import Any, Dict, List
+import jwt
+from jwt import PyJWKClient, DecodeError
+from firebase_admin import _utils
+
+_APP_CHECK_ATTRIBUTE = '_app_check'
+
+def _get_app_check_service(app) -> Any:
+    return _utils.get_app_service(app, _APP_CHECK_ATTRIBUTE, _AppCheckService)
+
+# should i accept an app (design doc doesn't have one) or just always make it none
+def verify_token(token: str, app=None) -> Dict[str, Any]:
+    """Verifies a Firebase App Check token.
+
+    Args:
+        token: A token from App Check.
+        app: An App instance (optional).
+
+    Returns:
+        Dict[str, Any]: A token's decoded claims
+        if the App Check token is valid; otherwise, a rejected promise..
+    """
+    return _get_app_check_service(app).verify_token(token)
+
+class _AppCheckService:
+    """Service class that implements Firebase App Check functionality."""
+
+    _APP_CHECK_GCP_API_URL = 'https://firebaseappcheck.googleapis.com'
+    _APP_CHECK_BETA_JWKS_RESOURCE = '/v1beta/jwks'
+
+    def __init__(self, app):
+        # the verification method should go in the service
+        project_id = app.project_id
+        if not project_id:
+            raise ValueError(
+                'Project ID is required to access App Check service. Either set the '
+                'projectId option, or use service account credentials. Alternatively, set the '
+                'GOOGLE_CLOUD_PROJECT environment variable.')
+        # Unsure what I should include in this constructor, or even if I should include one
+
+    @classmethod
+    def verify_token(cls, token: str) -> Dict[str, Any]:
+        """Verifies a Firebase App Check token."""
+        if token is None:
+            return None
+
+        # Obtain the Firebase App Check Public Keys
+        # Note: It is not recommended to hard code these keys as they rotate,
+        # but you should cache them for up to 6 hours.
+        url = f'{cls._APP_CHECK_GCP_API_URL}{cls._APP_CHECK_BETA_JWKS_RESOURCE}'
+        jwks_client = PyJWKClient(url)
+        signing_key = jwks_client.get_signing_key_from_jwt(token)
+
+        # Getting error "No value for argument 'header' in unbound
+        # method call (no-value-for-parameter)"
+        cls._has_valid_token_headers(jwt.get_unverified_header(token))
+        payload = cls._decode_and_verify(token, signing_key.key, "project_number")
+
+        # The token's subject will be the app ID, you may optionally filter against
+        # an allow list
+        return payload.get('sub')
+
+    def _has_valid_token_headers(self, header: Any) -> None:
+        """Checks whether the token has valid headers for App Check."""
+        # Ensure the token's header has type JWT
+        if header.get('typ') != 'JWT':
+            raise ValueError("The token received is not a JWT")
+        # Ensure the token's header uses the algorithm RS256
+        if header.get('alg') != 'RS256':
+            raise ValueError("JWT's algorithm does not have valid token headers")
```

Comment:
`The provided App Check token has incorrect algorithm. Expected RS256 but got ...`